### PR TITLE
Add JSON Web Toolkit Authentication for HTTP

### PR DESCRIPTION
**Pull request by rsto**
Adds JSON Web Toolkit Authentication to HTTP connections.

The "http_jwt_key_dir" option describes the feature:

    The absolute path to a directory containing one or more key files
    to authenticate JSON Web Tokens (RFC 7519) for HTTP connections.
    Keys for the following JWS algorithms are supported: "HS256",
    "HS384", "HS512", "RS256", "RS384", "RS512".

    A key file consists of one or more keys encoded in PEM format.
    RSA keys must be embedded between the lines
       "-----BEGIN PUBLIC KEY-----" and "-----END PUBLIC KEY-----"
    HMAC digest keys must be embedded between the lines
      "-----BEGIN HMAC KEY-----" and "-----END HMAC KEY-----",
    encoded in base64.
    Any lines before or after a PEM key definition are ignored,
    empty lines are ignored anywhere in the file.

    The JSON Web Token must be the value of the HTTP "Authorization" header,
    using the "Bearer" authentication scheme. The JWS Header must include the
    "alg" and "typ" parameter. A header with any other parameter is rejected.
    The JWS Payload must include the "sub" claim with the Cyrus user
    identifier as value. It may include the "iat" claim (see \fIhttp_jwt_max_age\fR).
    A payload with any other claim is rejected.

### CODE REVIEW

File: imap/http_jwt.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+{
+    is_enabled = 0;
+
+    if (!keystr) {
+        xsyslog(LOG_ERR, "Unexpected null key", NULL);
+        return HTTP_SERVER_ERROR;
+    }
+
+    if (strncmp(keystr, "HS256:", 6)) {
+        xsyslog(LOG_ERR, "Unexpected key algo specifier", NULL);
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_jwt.c:94]: Right, definitely should do that. The logging messages are a left-over from when we couldn't trust the data. But first, this doesn't apply to the init function anyway, and secondly we now first check the HMAC of authentication tokens before we decode JSON. So it should be rather safe to pass invalid or unexpected JSON entries to syslog.

File: imap/http_jwt.c [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+{
+    return ((c >= '0' && c <= '9') ||
+            (c >= 'a' && c <= 'z') ||
+            (c >= 'A' && c <= 'Z') ||
+            (c == '-' || c == '_'));
+}
+
+HIDDEN int http_jwt_init(const char *keystr, int age)
+{
+    is_enabled = 0;
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_jwt.c:86]: yes, thanks

File: imap/http_jwt.c [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            }
+        }
+    }
+    else if (max_age) {
+        xsyslog(LOG_ERR, "Missing \"iat\" claim", NULL);
+        goto done;
+    }
+
+    size_t sublen = strlen(sub);
+    if (sublen >= outlen) {
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_jwt.c:271]: yes, that's fine. We'd rather support one byte less than relying on the caller providing an extra byte.

File: imap/http_jwt.c [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+    return 1;
+}
+
+static int validate_signature(struct jwt *jwt)
+{
+    buf_reset(&jwt->buf);
+
+    const char *k = buf_base(&key);
+    int klen = buf_len(&key);
```
[the remark this answers is not in the document]
Comment (rsto) [reply imap/http_jwt.c:177]: yes, good catch. HMAC accepts a zero length key, so I added an assert

File: imap/http_jwt.c [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    if (charset_decode(&key, keystr + 6, strlen(keystr) - 6, ENCODING_BASE64URL)) {
+        xsyslog(LOG_ERR, "Invalid base64url key", "key=<%s>", keystr);
+        return HTTP_SERVER_ERROR;
+    }
+
+    static size_t HS256_MINLEN = 32; // see RFC 2104, section 3
+    static size_t HS256_MAXLEN = 64;
+
+    if (buf_len(&key) < HS256_MINLEN || buf_len(&key) > HS256_MAXLEN) {
+        xsyslog(LOG_ERR, "Key length outside allowed range", "range=[%zu:%zu]",
```
Comment (wolfsage) [inline imap/http_jwt.c:116]: It'd be nice to log the len we got, but we can live without it since it's in the config

**Review by ksmurchison — APPROVED** [review]
I haven't read RFC 7519 enough closely op verify that our JWT implementation is correct, but the changes to httpd and charset encoding LGTM.

**Comment by rsto** [comment]
Updated with additional review from @wolfsage

File: lib/charset.c [hunk 6]
```diff
@@ -443,10 +478,13 @@ static void qp2byte(struct convert_rock *rock, uint32_t c)
 static void b64_2byte(struct convert_rock *rock, uint32_t c)
 {
     struct b64_state *s = (struct b64_state *)rock->state;
-    char b = CHAR64(c);
+    char b = CHAR64(c, s->index);
 
-    /* could just be whitespace, ignore it */
-    if (b == XX) return;
+    if (b >= XS) {
```
Comment (brong) [inline lib/charset.c:483]: This is changing the base64 parsing behaviour to treat everything other than specific whitespace as `invalid`, which is probably fine but it does mean that all our base64 parsing is going to abort on stuff that it previously accepted.  I would want to make this a 'syslog but allow' at least for now, for non-URL base64.

File: lib/imapoptions [hunk 7]
```diff
@@ -1075,6 +1075,43 @@ Blank lines and lines beginning with ``#'' are ignored.
...
+
+{ "http_jwt_max_age", "0s", DURATION, "UNRELEASED" }
+/* Defines the timespan in seconds in which a JSON Web Token is valid
+   (see \fIhttp_jwt_key\fR). The value must be zero or positive.
+
+   If non-zero, the timespan starts at the point in time defined by the
+   "iat" claim of the JWS Payload and ends after the amount of seconds
+   defined in this option has passed. Tokens without an "iat" claim,
+   or with an issue date in the future, are rejected. There is no leeway
+   for clock skew.
```
[the remark this answers is not in the document]
Comment (rsto) [reply lib/imapoptions:1112]: Thanks. Updated.

**Comment by rsto** [comment]
Updated with RSA support as well, as well as SHA-384 and SHA-512 in addition to SHA-256. The configuration of keys now is file-based and allows to configure multiple keys. At the moment, only one configuration file is supported, so people have to maintain their keys in that file. Alternatively, we could also allow multiple key files to be read from a directory.

Embedding the HMAC keys in PEM with the line `-----BEGIN HMAC KEY-----" is non-standard, I haven't found a standard how to embed HMAC keys in PEM. Probably that should read `-----BEGIN HMAC PRIVATE KEY-----`, but there is no public key for HMAC.

**Comment by rouazana** [comment]
Nice, thank you very much!

cc @remk @chibenwa

**Review by brong — APPROVED** [review]
This is great.  Thanks for doing this rsto.

**Comment by rsto** [comment]
@elliefm I added a Cassandane test that updates the `TestCase.tear_down` routine. Could you please have a look if you agree with the change? https://github.com/cyrusimap/cyrus-imapd/pull/3708/commits/bda3849f61300ece5c62851f0a3dfa9a4c59ff60

**Comment by elliefm** [comment]
> @elliefm I added a Cassandane test that updates the `TestCase.tear_down` routine. Could you please have a look if you agree with the change? [bda3849](https://github.com/cyrusimap/cyrus-imapd/commit/bda3849f61300ece5c62851f0a3dfa9a4c59ff60)

Thanks for the ping!  I haven't tested it, but I assume you have and it works okay?

I wonder if there's a reason that you didn't hook into the existing certificate stuff.  Maybe the HMAC certificate is fundamentally different in some way I don't understand?  Or maybe you just didn't know it was there!

Have a look at:

* `cassandane/data/certs/` -- you could commit a new static file here, instead of hardcoded data and then needing tempfile to put it on disk
* `sub install_tls_certificates` in Cassandane::Instance to copy the static file into the instance's conf directory, which would eliminate the need for special cleanup in `tear_down`, because Cassandane won't be creating files outside of its own environment.  `install_tls_certificates` could be simply renamed to `install_certificates` if it's going to install non-tls certificates too.
* `magic(TLS => ...` in Cassandane::Cyrus::TestCase, as an example of how you could implement the `:HttpJWTAuthRSA` magic in terms of `install_tls_certificates`

---

If you've already looked at the above and it wasn't suitable for some reason, that's fine.  In that case:

The cleanup you've added to `tear_down` assumes there's only ever one Instance (`$self->{instance}`), and while I think this is true for the tests that use `:HttpJWTAuthRSA` at the moment, it won't be true if we start wanting to test this functionality in a replication or murder environment.

If we proceed with this implementation, I would like to see the cleanup moved into Cassandane::Instance's `sub cleanup` (which is already called from the line immediately below where you currently have it, but is also called correctly for every other instance that might have been instantiated).

I'm not sure it actually matters right now, since in this implementation only one tempfile will be created per test no matter how many instances that test starts up (they will all be configured to look at the same file), so it only needs to be cleaned up once.  But it doesn't hurt to unlink a file that was already unlinked, and moving it into `cleanup` will keep the instance cleanup in one place, which will be easier to think about later.

I've got a couple of style nits too, I'll add those as "suggestions" in a review comment.

File: cassandane/Cassandane/Cyrus/TestCase.pm [hunk 8]
```diff
@@ -889,6 +910,13 @@ sub tear_down
     if (defined $self->{instance})
     {
         eval { push @stop_errors, $self->{instance}->stop() };
+
+        # Clean up files outside basedir
+        my @delfiles = grep defined, (
+            $self->{instance}->{config}->get('http_jwt_key_path'),
+        );
```
Comment (elliefm) [inline cassandane/Cassandane/Cyrus/TestCase.pm:917]:
```suggestion
        my @delfiles = grep { defined } (
            $self->{instance}->{config}->get('http_jwt_key_path'),
        );
```

It's unusual for me to see the `grep EXPR, LIST` version of grep.  I think the `grep BLOCK LIST` version is generally preferred, though if there was a good reason for this I have long forgotten it!

I think if I personally were going to use the `grep EXPR, LIST` version for some reason (that I can't imagine), I would put the arguments in parens, c-style, something like `grep(defined, $self->{instance}->config}...);`, to make it more visually apparent to my future self that something "unusual" was happening.

Looks like, in Cassandane, we do use a bit of everything, but strongly prefer BLOCK:

```
ellie@debian:master:~/fastmail/cyrus-imapd$ack -ch 'grep \{' cassandane/Cassandane/
181
ellie@debian:master:~/fastmail/cyrus-imapd$ack -ch 'grep [^\{]' cassandane/Cassandane/
4
ellie@debian:master:~/fastmail/cyrus-imapd$ack -ch 'grep\(' cassandane/Cassandane/
1
```

**Review by elliefm — COMMENTED** [review]
Cassandane perl nits as promised ;)

I haven't looked at the Cyrus code at all, since I'm not familiar with the problem being solved so it would take a while, and I see Bron and Ken have already done so and approved it anyway.  Happy to take a look if you want though, lmk.

File: cassandane/Cassandane/Cyrus/TestCase.pm [hunk 9]
```diff
@@ -44,6 +44,7 @@ use attributes;
 use Data::Dumper;
 use Scalar::Util qw(refaddr);
 use List::Util qw(uniq);
+use File::Temp qw/ tempfile tempdir /;
```
[the remark this answers is not in the document]
Comment (rsto) [reply cassandane/Cassandane/Cyrus/TestCase.pm]: Thanks, I removed that tempfile approach completely

**Comment by rsto** [comment]
Thanks @elliefm . I did not know about data/certs. That's a much saner approach.

**Review by elliefm — APPROVED** [review]
Updated Cassandane looks good, and a full run passes.

Have not reviewed the C code, but giving my approval on the basis that that's already been reviewed and approved, and I'm approving the Cassandane changes.

**Comment by rsto** [comment]
@rouazana I now merged the JSON Web Token authentication patch on the main branch. Please check it out and if you find any issues or preferred changes, please let me know.

**Pull request merged**